enhancements/cert-manager/istio-csr-controller.md

```diff
@@ -0,0 +1,851 @@
+  and controller should cleanup all resources created for the `istio-csr` deployment.
+- As an OpenShift user, I should be able to install `istio-csr` in the upgrade cluster where istiod control plane
+  is active and should be able to update the certificate endpoint to `istio-csr` endpoint.
+- As an OpenShift user, I want to have an option to dynamically enable monitoring for the `istio-csr` project and
+  to use the OpenShift monitoring solution when required.
+
+### Goals
+
+- `cert-manager-operator` to be extended to manage `istio-csr` along with currently managed `cert-manager`.
+- New custom resource `istiocsr` to be made available to enable, configure and disable deployment.
+
+### Non-Goals
+
+- `istio-csr` can be used only with supported version of `OpenShift Service Mesh`. Please refer `Version Skew Strategy`
+  section for more details.
+
+## Proposal
+
+`istio-csr` will be managed by `cert-manager-operator` which can be enabled and configured by a user as and
+when required, using new custom resource which will be made available for configuring `istio-csr`.
+
+A new controller will be added to `cert-manager-operator` to manage and maintain the `istio-csr` deployment in
+desired state which can be monitored by user through the status sub-resource of the new custom resource.
+
+`istio-csr` controller will make use of static manifest templates for creating below resources required for
+successfully deploying `istio-csr` agent. Please refer `Implementation Details/Notes/Constraints` section for more details.
+* `certificate.cert-manager.io`
+* `clusterissuer.cert-manager.io`
+* `issuer.cert-manager.io`
+* `clusterrole.rbac.authorization.k8s.io`
+* `clusterrolebinding.rbac.authorization.k8s.io`
+* `deployment`
+* `service`
+* `servicemonitor.monitoring.coreos.com`
+* `namespace`
+* `role.rbac.authorization.k8s.io`
+* `rolebinding.rbac.authorization.k8s.io`
+* `serviceaccount`
+
+`istio-csr` CR object will be a singleton object, must be named `default` and an OpenShift user will have to
+create the same for enabling the deployment. `istio-csr` will be deployed in the namespace where the CR object is created.
+
+When `istiocsr` custom resource is created without defining `issuerref`, the operator will create a `ClusterIssuer` in the
+`cert-manager` namespace using a self-signed `Issuer`, which will act as a root CA. And the `ClusterIssuer` will be used to
+create an `Issuer` an intermediate CA, in the istio-csr deployed namespace and will used to obtain certificates for gRPC
+service made available for `OpenShift Service Mesh` and also for issuing the certificates requested by `OpenShift Service
+Mesh`. The `Issuer` in the istio-csr deployed namespace will be used for obtaining the signed certificates requested by 
+the `OpenShift Service Mesh`.
+
+Configurations made available in the spec of `istiocsr` custom resource are passed as command line arguments to `istio-csr`
```

Comment:
The configuration updates will make the system enter into an inconsistent state, for example, updates to certificate specfic configurations like issuer, trust bundles (CA configurations), algorithm updates. The configuration updates should only be allowed to low-impact fields. Can you please add more on how the operator intends to validate this flow?